### PR TITLE
Remove the single-symbol assumption in ConflictResolver Session

### PR DESCRIPTION
**Pull request by Cosifne**
This is the second milestone to support renaming multiple symbols. This is more likely to be a refactoring, and it doesn't introduce any new function.
The main changes of this PR are
1. For all the places using `ConflictResolution`, they are changed to support multiple symbols.
2. `ConflictSolution.Session` is changed to an abstract class. All the 'Single rename symbol' assumptions are removed there.
There is one check based on the 'Single Rename symbol' assuption. https://sourceroslyn.io/#Microsoft.CodeAnalysis.Workspaces/Rename/ConflictEngine/ConflictResolver.Session.cs,620
And I have put it in the subclass.
3. I introduce a 'DocumentRenameInfo' class to store all the rename data.

****
The next step of this PR, is to add `MultipleSymbolsRenameSession`, which handles the special conflict resolution logic for multiple symbols (and add entry in renamer and unit tests)

### CODE REVIEW

**Comment by Cosifne** [comment]
Perf result after:
image
Before:
image

When running benchmark here https://github.com/dotnet/roslyn/pull/63246

File: src/Workspaces/CSharp/Portable/Rename/CSharpRenameRewriterLanguageService.cs [hunk 1]
```diff
@@ -123,13 +123,7 @@ public RenameRewriter(RenameRewriterParameters parameters) : base(visitIntoStruc
                 _isProcessingComplexifiedSpans = false;
                 _skipRenameForComplexification = 0;
 
-                // TODO: These contexts are not changed for a document. ConflictResolver.Session should be refactored to cache them in a dictionary,
```
Comment (Cosifne) [inline src/Workspaces/CSharp/Portable/Rename/CSharpRenameRewriterLanguageService.cs:126]: All these data do not change during the renaming. So they are all cached in DocumentRenameInfo

File: src/Workspaces/Core/Portable/Rename/ConflictEngine/ConflictResolver.Session.cs [hunk 2]
```diff
@@ -613,16 +660,11 @@ private async Task DebugVerifyNoErrorsAsync(MutableConflictResolution conflictRe
                             }
                             else
                             {
-                                var newMetadataName = symbol.ToDisplayString(s_metadataSymbolDisplayFormat);
-                                var oldMetadataName = conflictAnnotation.RenameDeclarationLocationReferences[symbolIndex].Name;
-                                if (newLocation == null ||
-                                    newLocation.IsInSource ||
-                                    !HeuristicMetadataNameEquivalenceCheck(
```
Comment (Cosifne) [inline src/Workspaces/Core/Portable/Rename/ConflictEngine/ConflictResolver.Session.cs:620]: This is the only conflict check I don't think can expand to the multiple symbol rename scenario.

File: src/Workspaces/Core/Portable/Rename/ConflictEngine/ConflictResolver.Session.cs [hunk 3]
```diff
@@ -304,15 +346,13 @@ private async Task DebugVerifyNoErrorsAsync(MutableConflictResolution conflictRe
             {
                 try
                 {
-                    _documentOfRenameSymbolHasBeenRenamed |= documentIdsForConflictResolution.Contains(_documentIdOfRenameSymbolDeclaration);
```
Comment (Cosifne) [inline src/Workspaces/Core/Portable/Rename/ConflictEngine/ConflictResolver.Session.cs:307]: We don't need to store a separate bool value, because `RenameSpansTracker` stores all the changed document Ids during the renaming process.

File: src/Workspaces/Core/Portable/Rename/ConflictEngine/ConflictResolver.Session.cs [hunk 4]
```diff
@@ -686,63 +731,84 @@ private async Task<ISymbol> GetRenamedSymbolInCurrentSolutionAsync(MutableConfli
             /// The method determines the set of documents that need to be processed for Rename and also determines
             ///  the possible set of names that need to be checked for conflicts.
             /// </summary>
-            private async Task FindDocumentsAndPossibleNameConflictsAsync()
```
Comment (Cosifne) [inline src/Workspaces/Core/Portable/Rename/ConflictEngine/ConflictResolver.Session.cs:689]: There are no real change in this function. Only make it static

File: src/Workspaces/Core/Portable/Rename/SymbolicRenameLocations.cs [hunk 5]
```diff
@@ -26,7 +26,6 @@ internal sealed partial class SymbolicRenameLocations
         public readonly Solution Solution;
         public readonly ISymbol Symbol;
         public readonly SymbolRenameOptions Options;
-        public readonly CodeCleanupOptionsProvider FallbackOptions;
```
Comment (Cosifne) [inline src/Workspaces/Core/Portable/Rename/SymbolicRenameLocations.cs:29]: It is removed from the RenameLocations because I think one rename symbol should map to one `SymbolicRenameLocations`, so if we have multiple rename symbols, then we should have multiple `SymbolicRenameLocations` + one CodeCleanUpOptionsProvider.

File: src/Workspaces/Core/Portable/Rename/ConflictEngine/ConflictResolver.Session.cs [hunk 6]
```diff
@@ -28,56 +27,80 @@ internal static partial class ConflictResolver
...
+                _symbolToReplacementTextValid = symbolToReplacementTextValid;
+                _symbolToDeclarationDocumentAndLocation = symbolToDeclarationDocumentAndLocation;
             }
 
-            private SymbolRenameOptions RenameOptions => _renameLocationSet.Options;
-            private CodeCleanupOptionsProvider FallbackOptions => _renameLocationSet.FallbackOptions;
+            /// <summary>
+            /// Whether the <param name="newReferencedSymbol"/> has conflict with <param name="renameDeclarationLocationReference"/>
+            /// </summary>
+            protected abstract bool HasConflictForMetadataReference(RenameDeclarationLocationReference renameDeclarationLocationReference, ISymbol newReferencedSymbol);
```
Comment (Cosifne) [inline src/Workspaces/Core/Portable/Rename/ConflictEngine/ConflictResolver.Session.cs:101]: The only thing based on the 'Single Symbol' rename assumption.
I extract the implementation logic to 'SingleSymbolRenameSession'.
And to rename 'multiple symbols', we just need to create a derived type from 'Session', and implement this function.

File: src/EditorFeatures/Core/InlineRename/AbstractEditorInlineRenameService.InlineRenameReplacementInfo.cs [hunk 7]
```diff
@@ -15,17 +15,20 @@ private class InlineRenameReplacementInfo : IInlineRenameReplacementInfo
...
                 _conflicts = conflicts;
+                ReplacementTextValid = _conflicts.SymbolToReplacementTextValid[renameSymbol];
             }
 
             public IEnumerable<DocumentId> DocumentIds => _conflicts.DocumentIds;
 
             public Solution NewSolution => _conflicts.NewSolution!;
 
-            public bool ReplacementTextValid => _conflicts.ReplacementTextValid!.Value;
+            public bool ReplacementTextValid { get; }
```
Comment (ryzngard) [inline src/EditorFeatures/Core/InlineRename/AbstractEditorInlineRenameService.InlineRenameReplacementInfo.cs:31]:
```suggestion
            public bool ReplacementTextValid => _conflicts.SymbolToReplacementTextValid[renameSymbol];
```

File: src/EditorFeatures/Test/Rename/RenameRewriterTests.Verifier.cs [hunk 8]
```diff
@@ -129,36 +126,35 @@ public Verifier(string workspaceXml)
...
+                            }
 
                             var stringAndCommentsRenameContexts = locationsInDocument
                                 .WhereAsArray(location => location.IsRenameInStringOrComment)
-                                .SelectAsArray(location => new LocationRenameContext(location, symbolContext));
-                            stringAndCommentsContextsBuilder.AddRange(stringAndCommentsRenameContexts);
+                                .SelectAsArray(location => new StringAndCommentRenameContext(location, replacementText));
+                            foreach (var stringAndCommentRenameContext in stringAndCommentsRenameContexts)
+                            {
+                                documentRenameBuilder.AddStringAndCommentRenameContext(stringAndCommentRenameContext);
```
Comment (ryzngard) [inline src/EditorFeatures/Test/Rename/RenameRewriterTests.Verifier.cs:156]: nit: Why not just have a range overload for this? Makes the code slightly cleaner here

File: src/Workspaces/CSharp/Portable/Rename/CSharpRenameRewriterLanguageService.cs [hunk 9]
```diff
@@ -123,13 +123,7 @@ public RenameRewriter(RenameRewriterParameters parameters) : base(visitIntoStruc
...
-                _renamedSymbolContexts = CreateSymbolKeyToRenamedSymbolContextMap(parameters.RenameSymbolContexts, SymbolKey.GetComparer());
-                _textSpanToLocationContextMap = CreateTextSpanToLocationContextMap(parameters.TokenTextSpanRenameContexts);
-                _stringAndCommentRenameContexts = GroupStringAndCommentsTextSpanRenameContexts(parameters.StringAndCommentsTextSpanRenameContexts);
-                _replacementTexts = _renamedSymbolContexts.Select(pair => pair.Value.ReplacementText).ToImmutableHashSet();
-                _originalTexts = _renamedSymbolContexts.Select(pair => pair.Value.OriginalText).ToImmutableHashSet();
-                _allPossibleConflictNames = _renamedSymbolContexts.SelectMany(pair => pair.Value.PossibleNameConflicts).ToImmutableHashSet();
+                (_textSpanToLocationContextMap, _renamedSymbolContexts, _stringAndCommentRenameContexts, _replacementTexts, _originalTexts, _allPossibleConflictNames) = parameters.DocumentRenameInfo;
```
Comment (ryzngard) [inline src/Workspaces/CSharp/Portable/Rename/CSharpRenameRewriterLanguageService.cs:126]: this is a quite heavy handed expansion :) 

Should we just hold onto `_documentRenameInfo` as the field?

File: src/Workspaces/Core/Portable/Microsoft.CodeAnalysis.Workspaces.csproj [hunk 10]
```diff
@@ -114,7 +114,7 @@
     <InternalsVisibleTo Include="Microsoft.CodeAnalysis.CSharp.Workspaces.UnitTests" />
     <InternalsVisibleTo Include="Microsoft.CodeAnalysis.VisualBasic.Workspaces.UnitTests" />
     <InternalsVisibleTo Include="Microsoft.VisualStudio.LanguageServices.CSharp.UnitTests" />
-    <InternalsVisibleTo Include="Microsoft.CodeAnalysis.LanguageServerIndexFormat.Generator.UnitTests"/>
+    <InternalsVisibleTo Include="Microsoft.CodeAnalysis.LanguageServerIndexFormat.Generator.UnitTests" />
```
Comment (ryzngard) [inline src/Workspaces/Core/Portable/Microsoft.CodeAnalysis.Workspaces.csproj:117]: I was going to say revert, but I'm guessing this is wrong without the space

File: src/Workspaces/Core/Portable/Rename/ConflictEngine/ConflictResolver.SingleSymbolRenameSession.cs [hunk 11]
```diff
@@ -0,0 +1,137 @@
...
+                var documentIdOfRenameSymbolDeclaration = renameLocationSet.Solution.GetRequiredDocument(renameSymbolDeclarationLocation.SourceTree!).Id;
+
+                using var _ = PooledDictionary<DocumentId, DocumentRenameInfo>.GetInstance(out var documentIdToRenameInfoBuilder);
+                var (documentsIdsToBeCheckedForConflict, possibleNameConflicts) = await FindDocumentsAndPossibleNameConflictsAsync(
+                    renameLocationSet,
+                    replacementText,
+                    originalText,
+                    cancellationToken).ConfigureAwait(false);
+
+                var replacementTextValid = IsIdentifierValid_Worker(
```
Comment (ryzngard) [inline src/Workspaces/Core/Portable/Rename/ConflictEngine/ConflictResolver.SingleSymbolRenameSession.cs:45]: Can we move `IsIdentifierValid_Worker` to this class? I don't think it's called elsewhere

File: src/Workspaces/Core/Portable/Rename/ConflictEngine/ConflictResolver.SingleSymbolRenameSession.cs [hunk 12]
```diff
@@ -0,0 +1,137 @@
...
+                    renameLocationSet.Solution,
+                    replacementText,
+                    documentsIdsToBeCheckedForConflict.Select(documentId => documentId.ProjectId).Distinct());
+
+                var documentIdToRenameLocations = renameLocations
+                    .GroupBy(location => location.DocumentId)
+                    .ToDictionary(grouping => grouping.Key);
```
Comment (ryzngard) [inline src/Workspaces/Core/Portable/Rename/ConflictEngine/ConflictResolver.SingleSymbolRenameSession.cs:52]:
```suggestion
                    .ToDictionary(grouping => grouping.Key);
                    
```

File: src/Workspaces/Core/Portable/Rename/ConflictEngine/ConflictResolver.SingleSymbolRenameSession.cs [hunk 13]
```diff
@@ -0,0 +1,137 @@
...
+                foreach (var documentId in documentsIdsToBeCheckedForConflict)
+                {
+                    using var documentRenameInfoBuilder = new DocumentRenameInfo.Builder();
+
+                    // Conflict checking documents is a superset of the rename locations. In case this document is not a documents of rename locations,
+                    // just passing an empty rename information to check for conflicts.
+                    var renameLocationsInDocument = documentIdToRenameLocations.ContainsKey(documentId)
```
Comment (ryzngard) [inline src/Workspaces/Core/Portable/Rename/ConflictEngine/ConflictResolver.SingleSymbolRenameSession.cs:59]: This is a little hard to follow. Reading below, it looks like the only work left is 

```c#
                    documentRenameInfoBuilder.AddRenamedSymbol(symbol, replacementText, replacementTextValid, possibleNameConflicts);
                    documentIdToRenameInfoBuilder[documentId] = documentRenameInfoBuilder.ToRenameInfo();
```

Everything else is skipped because of the empty array. Can we restructure this to clarify more? Or move to another method

**Review by ryzngard — APPROVED** [review]

File: src/Workspaces/CSharp/Portable/Rename/CSharpRenameRewriterLanguageService.cs (hunk 9 above)
Comment (Cosifne) [reply src/Workspaces/CSharp/Portable/Rename/CSharpRenameRewriterLanguageService.cs:126]: It's the one of my favorite parts of the record type : )
But I agree they can be fields

File: src/Workspaces/Core/Portable/Rename/ConflictEngine/ConflictResolver.SingleSymbolRenameSession.cs (hunk 11 above)
Comment (Cosifne) [reply src/Workspaces/Core/Portable/Rename/ConflictEngine/ConflictResolver.SingleSymbolRenameSession.cs:45]: Yeah I can put it here in this PR, but in my next PR I need to move it back to the base type again : )

File: src/Workspaces/Core/Portable/Rename/ConflictEngine/ConflictResolver.SingleSymbolRenameSession.cs (hunk 13 above)
Comment (Cosifne) [reply src/Workspaces/Core/Portable/Rename/ConflictEngine/ConflictResolver.SingleSymbolRenameSession.cs:59]: Yup this is a pretty tricky part of the conflict resolution. (I was thinking like this at the beginning)
Basically `ConflictCheckingDocument` = `All the renaming documents` + `All the possible naming conflict documents`, the second part contains some documents which could just have the same text in some of the identifiers. And we need to visit them to annotate the potential conflict locations.

Let me refactor this part a little bit to make it more clear
